lib/mini_racer/truffleruby.rb

```diff
@@ -0,0 +1,353 @@
+      end
+      if snapshot
+        @snapshot = snapshot
+      elsif isolate
+        @snapshot = isolate.instance_variable_get(:@snapshot)
+      else
+        @snapshot = nil
+      end
+    end
+
+    def dispose_unsafe
+      @context.close
+    end
+
+    def eval_unsafe(str, filename)
+      @entered = true
+      if !@has_entered && @snapshot
+        snapshot_src = encode(@snapshot.instance_variable_get(:@source))
+        begin
+          eval_in_context(snapshot_src, filename)
+        rescue RuntimeError => e
+          if e.message == "Polyglot::InnerContext was terminated forcefully"
+            raise ScriptTerminatedError, "JavaScript was terminated (either by timeout or explicitly)"
+          else
+            raise e
+          end
+        end
+      end
+      @has_entered = true
+      raise RuntimeError, "TruffleRuby does not support eval after stop" if @stopped
+      raise ArgumentError, "wrong type argument #{str.class} (should be a string)" unless str.kind_of?(String)
+      raise ArgumentError, "wrong type argument #{filename.class} (should be a string)" unless filename.nil? || filename.kind_of?(String)
+
+      str = encode(str)
+      begin
+        translate do
+          eval_in_context(str, filename)
+        end
+      rescue RuntimeError => e
+        if @current_exception
+          e = @current_exception
+          @current_exception = nil
+          raise e
+        elsif e.message == "Polyglot::InnerContext was terminated forcefully"
+          raise ScriptTerminatedError, "JavaScript was terminated (either by timeout or explicitly)"
+        else
+          raise e, e.message
+        end
+      rescue SystemStackError => e
+        raise RuntimeError, e.message
+      end
+    ensure
+      @entered = false
+    end
+
+    def call_unsafe(function_name, *arguments)
+      @entered = true
+      if !@has_entered && @snapshot
+        src = encode(@snapshot.instance_variable_get(:source))
+        begin
+          eval_in_context(src)
+        rescue RuntimeError => e
+          raise e unless e.message == "Polyglot::InnerContext was terminated forcefully"
+        end
+      end
+      @has_entered = true
+      raise RuntimeError, "TruffleRuby does not support call after stop" if @stopped
+      begin
+        translate do
+          function = eval_in_context(function_name)
+          function.call(*convert_ruby_to_js(arguments))
+        end
+      rescue RuntimeError => e
+        raise e unless e.message == "Polyglot::InnerContext was terminated forcefully"
+      rescue SystemStackError => e
+        raise RuntimeError, e.message
+      end
+    ensure
+      @entered = false
+    end
+
+    def create_isolate_value
+      # Returning a dummy object since TruffleRuby does not have a 1-1 concept with isolate.
+      # However, code and ASTs are shared between contexts.
+      Isolate.new
+    end
+
+    def isolate_mutex
+      @isolate_mutex
+    end
+
+    class ExternalFunction
```

Comment:
Maybe move this at the top of bottom, so not in the middle of methods

Reply:
Moved class to top